### ISSUE TITLE
No pick_value(pid(), ...) clause present

### ISSUE THREAD

**Issue by erszcz** [josefs/Gradualizer#398]
Another crash from #373:

```
apps/aecore/src/aec_peers.erl
escript: exception error: no function clause matching
                 gradualizer_lib:pick_value({type,0,pid,[]}, ...
```

**Comment by VincentVanBork** [comment]
Could I have look at it ?

**Comment by erszcz** [comment]
@VincentVanBork Sure, feel free to do so. It's not fixed yet. It's a missing clause in the aforementioned function.

**Comment by zuiderkwast** [comment]
So what's the idea? To return the absform of the function call `self()` or something else?

**Comment by erszcz** [comment]
I was thinking of a placeholder, e.g. `<0.0.0>`. Technically, it's not a valid pid itself, but I imagine it would be good enough. What do you think?

**Comment by zuiderkwast** [comment]
I think there is no abstract form for `<0.0.0>`.

**Comment by erszcz** [comment]
Hmm, yeah, indeed. This is trickier than it seemed to be :D

However, we know that the value returned from `pick_value` is only actually used for printing - we could abuse this knowledge a bit and return something which would result in `<0.0.0>` being printed, while technically not being the abstract form of a pid. WDYT?

Alternatively, we could make it explicit in code and just refactor the whole thing to `print_value` which would be free to return whatever printable string/binary would need to.

**Comment by VincentVanBork** [comment]
If I understand correctly the simplest solution would be to hardcode tuple with {pid,  erl_anno:new(0), '<0.0.0.>'} as a return in pick_value?

**Comment by zuiderkwast** [comment]
We use erl_pp:expr/1 to format it back to Erlang, so this doesn't work:

```Erlang
1> lists:flatten(erl_pp:expr({pid, erl_anno:new(0), '<0.0.0.>'})).      
"INVALID-FORM:{pid,0,'<0.0.0.>'}:"
```

Using `self()` would work though:

```Erlang
2> lists:flatten(erl_pp:expr({call, 0, {atom, 0, self}, []})).
"self()"
```

**Comment by erszcz** [comment]
Hmm, what about `pid(_,_,_)`?

```erlang
16> merl:quote("pid(0,0,0)").
{call,1,
      {atom,1,pid},
      [{integer,1,0},{integer,1,0},{integer,1,0}]}
17> merl:quote("pid(_,_,_)").
{call,1,{atom,1,pid},[{var,1,'_'},{var,1,'_'},{var,1,'_'}]}
19> io:format("~ts\n", [(erl_pp:expr(merl:quote("pid(_,_,_)")))]).
pid(_, _, _)
```

`self()` would work, but it suggests some specific pid, which might be misleading.

**Comment by zuiderkwast** [comment]
Sure, that works.

Nether pid(0,0,0) nor self() are values, but expressions that evaluate to values in a very strait-forward way, so the error message would be very easy to understand anyway.

pid(0,0,0) is very unusual to see in code though, while self() is often seen so it might be easier to understand. Is there any case where self() is wrong? Or pid(0,0,0) for that matter? I'm trying to think of something....

```Erlang
-spec f(pid()) -> ok.
f(P) ->
    Q = pid(0,0,0),
    case P of
        Q -> ok
    end.
```

Here, "pid(0,0,0) is not covered" would be incorrect.

**Comment by erszcz** [comment]
Indeed, but exactly the same example would be incorrect with `Q = self()`.

I think `pid(_,_,_)` with the wildcards could be interpreted as "any pid", which IMO makes sense. I'm not sure if we'll ever do refinement on the pid type, so `pick_value` will likely not need to ever return an exact pid value. And if it does, I think we can change it in the future. Thoughts?

**Comment by VincentVanBork** [comment]
The quickest and most clean in either cases seems like `lists:flatten(erl_pp:expr({value, erl_anno:new(0), self() })).` Could this be okay?

**Comment by erszcz** [comment]
I think literally

```
{call,1,{atom,1,pid},[{var,1,'_'},{var,1,'_'},{var,1,'_'}]}
```

or

```
{call,1,{atom,1,self},[]}
```

are the best options (in my personal order of preference).

**Comment by zuiderkwast** [comment]
`pid(_,_,_)` is a pattern. It's not a valid expression. Therefore, I think it's weird. I prefer self(). But, @erszcz you're right, the same example with `Q = self()` would be wrong too. Don't we have this problem with other types too, let's say floats? We pick 0.0 as the unmatched value even if `Q = 0.0` in a similar example with floats...

We'll never be able to do refinement on pids, since there are no subtypes of pid() (apart from none()). All we can do is to match all pids in a clause like `Pid when is_pid(Pid)` that let's us refine e.g. `integer() | pid()` to just `integer()`.

**Comment by zuiderkwast** [comment]
> The quickest and most clean in either cases seems like `lists:flatten(erl_pp:expr({value, erl_anno:new(0), self() })).` Could this be okay? Instead of `call` just put `value` and `self()` or `<0.0.0>`

@VincentVanBork Interesting. I didn't know this absform existed. It seems to work! Is it documented anywhere?

```Erlang
9> lists:flatten(erl_pp:expr({value, erl_anno:new(0), self() })).
"<0.82.0>"
```

**Comment by VincentVanBork** [comment]
> > The quickest and most clean in either cases seems like `lists:flatten(erl_pp:expr({value, erl_anno:new(0), self() })).` Could this be okay? Instead of `call` just put `value` and `self()` or `<0.0.0>`
> 
> @VincentVanBork Interesting. I didn't know this absform existed. It seems to work! Is it documented anywhere?
> 
> ```erlang
> 9> lists:flatten(erl_pp:expr({value, erl_anno:new(0), self() })).
> "<0.82.0>"
> ```
I don't think so I found it when looking at source code, cause it is more of a erl_pp problem than that of Gradualizer.
https://github.com/erlang/otp/blob/4dd776a44d5831484f5cef93dafed459b183034c/lib/stdlib/src/erl_pp.erl#L758

**Comment by zuiderkwast** [comment]
"%% Special case for straight values."

:-) Undocumented but OK, I think we can use it.

**Comment by erszcz** [comment]
Up until some point, I don't remember which version exactly, it was not possible to paste into the shell terms which the shell could itself print (pids, refs, and ports). This made data post-processing very inconvenient. For some time now it's possible:

```
1> self().
<0.83.0>
2> <0.83.0>. %% pasted
<0.83.0>
```

However, it's still not possible to paste `<0.83.0>` into an `.erl` file and have that parsed / compiled. This makes me think the call version, so `self()` or `pid(0,0,0)`, is preferable.

> Don't we have this problem with other types too, let's say floats? We pick -1.0 as the unmatched value even if Q = -1.0 in a similar example with floats...

Indeed, we do, it's all a bit arbitrary. The intention really is to print an example understandable to the user, not to be 100% accurate. That's why I think `pid(_,_,_)` is fine, as it gives the right impression. I get, though, that since it's a pattern it's a bit strange in this role, so I think `self()` or `pid(0,0,0)` would be less ambiguous and less strange.

**Comment by erszcz** [comment]
It seems `self()` is the least controversial for all of us, so maybe that's the one to use?

**Comment by VincentVanBork** [comment]
@erszcz for me yes I like that it is value and not pattern.

**Comment by zuiderkwast** [comment]
Actually I like <0.0.0> even more.  It is a value whish is printed and it doesn't matter thst you can't write it as a literal.

**Comment by zuiderkwast** [comment]
But I'm fine with self() too.

**Comment by erszcz** [comment]
Note to self: rechecking if that's still an issue with the current master branch would be insightful.

**Comment by zuiderkwast** [comment]
```Erlang
erlang:list_to_pid("<0.0.0>")
```
